File: crypto/hmac.go
package crypto

import (
	"crypto/hmac"
	"crypto/sha256"
	"errors"
)

const (
	// HMACFullSize is the size of a full HMAC-SHA256 output in bytes.
	HMACFullSize = sha256.Size // 32 bytes
)

var (
	// ErrInvalidKeySize is returned when a key has an incorrect length.
	ErrInvalidKeySize = errors.New("crypto: invalid key size")

	// ErrInvalidTruncation is returned when a truncation length exceeds the full HMAC size.
	ErrInvalidTruncation = errors.New("crypto: truncation exceeds HMAC size")
)

// HMACSHA256 computes the full HMAC-SHA256 of data using the given key.
// Returns a 32-byte MAC.
func HMACSHA256(key, data []byte) []byte {
	mac := hmac.New(sha256.New, key)
	mac.Write(data)
	return mac.Sum(nil)
}

// HMACSHA256Truncated computes HMAC-SHA256 and truncates the result to n bytes.
//
// The MyClerk Protocol uses n=8 for Tier 2 authentication. Truncation to fewer
// bytes reduces security but saves bandwidth — 8 bytes (64 bits) provides
// adequate integrity for the protocol's use case.
//
// Returns an error if n exceeds HMACFullSize (32 bytes) or is zero.
func HMACSHA256Truncated(key, data []byte, n int) ([]byte, error) {
	if n <= 0 || n > HMACFullSize {
		return nil, ErrInvalidTruncation
	}
	full := HMACSHA256(key, data)
	result := make([]byte, n)
	copy(result, full[:n])
	return result, nil
}

// VerifyHMACSHA256 verifies a full HMAC-SHA256 tag.
// Uses constant-time comparison to prevent timing attacks.
func VerifyHMACSHA256(key, data, expectedMAC []byte) bool {
	computed := HMACSHA256(key, data)
	return hmac.Equal(computed, expectedMAC)
}

// VerifyHMACSHA256Truncated verifies a truncated HMAC-SHA256 tag.
// The expectedMAC length determines the truncation size.
// Uses constant-time comparison to prevent timing attacks.
func VerifyHMACSHA256Truncated(key, data, expectedMAC []byte) bool {
	n := len(expectedMAC)
	if n <= 0 || n > HMACFullSize {
		return false
	}
	full := HMACSHA256(key, data)
	return hmac.Equal(full[:n], expectedMAC)
}

File: kt/src/main/kotlin/de/arcan/arclib/crypto/Hmac.kt
package de.arcan.arclib.crypto

import java.security.MessageDigest
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

object Hmac {
    fun sha256(key: ByteArray, data: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(data)
    }

    fun sha256Truncated(key: ByteArray, data: ByteArray, n: Int): ByteArray {
        require(n in 1..32) { "truncation must be 1-32 bytes" }
        return sha256(key, data).copyOfRange(0, n)
    }

    fun verifySha256(key: ByteArray, data: ByteArray, expectedMac: ByteArray): Boolean {
        val computed = sha256(key, data)
        return MessageDigest.isEqual(computed, expectedMac)
    }
}
